### PR TITLE
Delegate inspection API in the DBI

### PR DESCRIPTION
**Pull request by hoyosjs**
The changes in the DAC are expected to be 
There's some cases in GetDelegateType that are missing that I'm adding currently, as well as trimming different places and adding comments

Cases tested (On Windows):
- R2R methods in CoreLib
- Static Methods
- Member methods
- Generics

Known limitations:
- Will report fail on a native wrapper delegate, a wrapper delegate (only used on arm), and multi-function multicast delegates
- For generics, the non-instantiated function will be returned

cc @tommcdon

### CODE REVIEW

File: src/inc/cordebug.idl [hunk 1]
```diff
@@ -6491,6 +6491,33 @@ interface ICorDebugObjectValue2 : IUnknown
...
+]
+interface ICorDebugDelegateObjectValue : IUnknown
+{
+    /*
+     * GetTarget retrieves the object on which the delegate calls the function.
+     * Returns:
+     *  - S_OK:
+     *      - The ICorDebugObjectValue object targeted by the function.
+     *      - NULL if the delegate function is a static function.
+     *  - HRESULT XXX for multicast delegates.
```
Comment (noahfalk) [inline src/inc/cordebug.idl]: This should probably say 'failing HRESULT' or something more detailed.

File: src/vm/common.h [hunk 2]
```diff
@@ -153,6 +153,7 @@ typedef DPTR(class NDirectMethodDesc)   PTR_NDirectMethodDesc;
 typedef VPTR(class Thread)              PTR_Thread;
 typedef DPTR(class Object)              PTR_Object;
 typedef DPTR(PTR_Object)                PTR_PTR_Object;
+typedef DPTR(class DelegateObject)              PTR_DelegateObject;
```
Comment (noahfalk) [inline src/vm/common.h]: Nit: spacing

File: src/debug/daccess/dacdbiimpl.h [hunk 3]
```diff
@@ -349,6 +349,26 @@ class DacDbiInterfaceImpl :
     // Returns true if the argument is a runtime callable wrapper
     BOOL IsRcw(VMPTR_Object vmObject);
 
+    HRESULT GetMethodDescPtrFromIpEx(
```
Comment (noahfalk) [inline src/debug/daccess/dacdbiimpl.h]: I don't see any place where mscordbi calls this function. You could make this private and exclude from the interface.

File: src/debug/daccess/dacdbiimpl.cpp [hunk 4]
```diff
@@ -3406,6 +3406,187 @@ BOOL DacDbiInterfaceImpl::IsExceptionObject(MethodTable* pMT)
...
+    DD_ENTER_MAY_THROW;
+
+    if (vmObject.IsNull())
+        return FALSE;
+
+    Object *pObj = vmObject.GetDacPtr();
+    return pObj->GetGCSafeMethodTable()->IsDelegate();
+}
+
+HRESULT DacDbiInterfaceImpl::GetDelegateType(VMPTR_Object delegateObject, DelegateType *delegateType)
```
Comment (noahfalk) [inline src/debug/daccess/dacdbiimpl.cpp]: Ideally in the DAC code we can share the runtime code by refactoring it, but when that doesn't work well it is a good idea to add comments in the runtime warning that any change to that code needs to be replicated here, and vice versa. 
My suggestion at this point is that you add the comments, we are too far along and timeline is way too tight to consider the refactor.

File: src/debug/daccess/dacdbiimpl.cpp [hunk 5]
```diff
@@ -3406,6 +3406,187 @@ BOOL DacDbiInterfaceImpl::IsExceptionObject(MethodTable* pMT)
...
+        }
+    }
+    else
+    {
+        if (pInvocationList != NULL)
+        {
+            PTR_MethodTable invocationListMT = pInvocationList->GetGCSafeMethodTable();
+
+            if (invocationListMT->IsArray())
+                *delegateType = kTrueMulticastDelegate;
```
Comment (noahfalk) [inline src/debug/daccess/dacdbiimpl.cpp]: Nit: The name "True" multicast delegate implies there is some other not-so-true multicast delegate. I would either drop the "True" or make it clearer what this is being distinguished from.

**Review by noahfalk — APPROVED** [review]
This looks good to me. I commented on a few nits : )

**Comment by noahfalk** [comment]
> There's some cases in GetDelegateType that are missing that I'm adding currently

I think you are out of time for Preview 7. I would recommend:
a) If you are aware of any bugs which are severe enough that you don't think the feature should be shipped in Preview 7 as-is, let folks know ASAP (I'm assuming there aren't any)
b) If you are aware of issues that aren't so terrible, but you still think they deserve to be fixed - file an issue that can potentially be addressed before ZBB. I would not change a single line of code at this point other than to make CI pass (corollary - acting on my feedback above isn't required to make CI pass, so put it in the 'file an issue bucket')

File: src/debug/daccess/dacdbiimpl.cpp (hunk 4 above)
Comment (hoyosjs) [reply src/debug/daccess/dacdbiimpl.cpp]: Indeed, Added.

File: src/debug/daccess/dacdbiimpl.cpp (hunk 5 above)
Comment (hoyosjs) [reply src/debug/daccess/dacdbiimpl.cpp]: The name comes from `COMDelegate::IsTrueMulticastDelegate`. I will add an appropriate comment.

**Comment by hoyosjs** [comment]
> I think you are out of time for Preview 7. I would recommend:
> a) If you are aware of any bugs which are severe enough that you don't think the feature should be shipped in Preview 7 as-is, let folks know ASAP (I'm assuming there aren't any)
> b) If you are aware of issues that aren't so terrible, but you still think they deserve to be fixed - file an issue that can potentially be addressed before ZBB. I would not change a single line of code at this point other than to make CI pass (corollary - acting on my feedback above isn't required to make CI pass, so put it in the 'file an issue bucket')

I think I can fail elegantly enough, so I'll go for option `b`.

**Review by tommcdon — APPROVED** [review]

**Comment by hoyosjs** [comment]
@noahfalk if you have no other comments, I'll merge this once all ci passes

**Pull request merged**